File: 796-rotate-string/solution.cpp
#include <string>

class Solution {
public:
    bool rotateString(string s, string goal) {
        // If lengths differ, s can never be transformed into goal
        if (s.length() != goal.length()) {
            return false;
        }

        // Any rotation of s is a substring of s + s
        string doubled = s + s;

        // Check if goal exists within the doubled string
        // string::find returns string::npos if the substring is not found
        return doubled.find(goal) != string::npos;
    }
};

File: 796-rotate-string/solution.py
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        """
        Determines if s can become goal after some number of shifts.
        
        A rotation of s means goal must have the same length as s,
        and goal must be a substring of s concatenated with itself.
        """
        # Step 1: Check if lengths match. If not, rotation is impossible.
        if len(s) != len(goal):
            return False
        
        # Step 2: Check if goal is a substring of s + s.
        # Python's 'in' operator uses a highly optimized string matching algorithm.
        return goal in (s + s)